### PR TITLE
p11sak fixes and enhancements

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

File: usr/sbin/p11sak/p11sak.c [hunk 1]
```diff
@@ -8880,9 +8947,140 @@ static CK_RV p11sak_export_sym_key(const struct p11sak_objtype *keytype,
...
+
+        rc = keytype->import_asym_pkey(keytype, pkey, false,
+                                       &attrs, &num_attrs);
+    } else if (keytype->import_asym_pem_data != NULL) {
+        rc = keytype->import_asym_pem_data(keytype,
+                                           (unsigned char *)spki.pValue,
+                                           spki.ulValueLen, false,
+                                           &attrs, &num_attrs);
+    } else {
+        warnx("No support for extracting %s public key",keytype->name);
```
Comment (jschmidb) [inline usr/sbin/p11sak/p11sak.c:9020]: blank before keytype->name ?

File: usr/sbin/p11sak/p11sak.c [hunk 2]
```diff
@@ -8880,9 +8947,140 @@ static CK_RV p11sak_export_sym_key(const struct p11sak_objtype *keytype,
...
+            goto done;
+        }
+    }
+
+    if (opt_new_label != NULL) {
+        rc = add_attribute(CKA_LABEL, opt_new_label, strlen(opt_new_label),
+                           &attrs, &num_attrs);
+        if (rc != CKR_OK) {
+            warnx("Failed to add %s key attribute CKA_LABEL: 0x%lX: %s",
+                    keytype->name, rc, p11_get_ckr(rc));
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/sbin/p11sak/p11sak.c:8984]: fixed

File: usr/sbin/p11sak/p11sak.c [hunk 3]
```diff
@@ -9495,12 +9503,25 @@ static CK_RV init_pkcs11(const struct p11sak_cmd *command)
...
 
-    if (pin == NULL)
-        pin = getenv(PKCS11_USER_PIN_ENV_NAME);
-    if (opt_force_pin_prompt || pin == NULL)
-        pin = pin_prompt(&buf_user_pin, "Please enter user PIN: ");
-    if (pin == NULL)
-        return CKR_FUNCTION_FAILED;
+    if (opt_no_login) {
+        if (opt_pin != NULL) {
+            warnx("Option '-p'/'--pin' is not allowed with '-N'/'--no-login'");
```
Comment (jschmidb) [inline usr/sbin/p11sak/p11sak.c:9839]: OK, here we may have the case that the pin ist set via env variable, but let's just silently ignore it.

File: usr/sbin/p11sak/p11sak.c [hunk 4]
```diff
@@ -9513,12 +9526,18 @@ static CK_RV init_pkcs11(const struct p11sak_cmd *command)
                   "'-N'/'--no-login'");
             return CKR_ARGUMENTS_BAD;
         }
+        if (opt_so) {
+            warnx("Option ' --so' is not allowed with '-N'/'--no-login'");
```
Comment (jschmidb) [inline usr/sbin/p11sak/p11sak.c:9848]: Remove blank in ' --so' ?

File: man/man1/p11sak.1.in [hunk 5]
```diff
@@ -1859,7 +1859,12 @@ is used. If
 .B PKCS11_SO_PIN
 is not set, then you will be prompted for the SO PIN. SO can only access public
 token objects
-.RB ( CKA_PRIVATE = FALSE ).
+.RB ( CKA_PRIVATE = FALSE ),
+but SO can set attribute
+.B CKA_TRUSTED
+to
+.B TRUE
+for secret key, public key, and certificate objects.
```
Comment (jschmidb) [inline man/man1/p11sak.1.in:1953]: Hmm ... also plural for keys? Like "... for secret keys, public keys, and certificate objects' ?

File: usr/sbin/p11sak/p11sak.c [hunk 6]
```diff
@@ -9288,6 +9559,36 @@ static CK_RV p11sak_export_key(void)
...
+        keytype = opt_keytype->private.ptr;
+
+    data.export_all = opt_force;
+
+    rc = iterate_objects(keytype, opt_label, opt_id, opt_attr,
+                         OBJCLASS_KEY, NULL,
+                         handle_key_pubkey_extract, &data);
+    if (rc != CKR_OK) {
+        warnx("Failed to iterate over key objects for type %s: 0x%lX: %s",
+                keytype != NULL ? keytype->name : "All", rc, p11_get_ckr(rc));
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/sbin/p11sak/p11sak.c:9578]: fixed

File: usr/sbin/p11sak/p11sak.c (hunk 4 above)
Comment (ifranzki) [reply usr/sbin/p11sak/p11sak.c:9848]: fixed (also a few lines above: `warnx("Option ' --force-pin-prompt' is not allowed with "` ).

File: man/man1/p11sak.1.in (hunk 5 above)
Comment (ifranzki) [reply man/man1/p11sak.1.in:1953]: hm, well, the 'objects' was meant for all 3 of them....
Not sure how to make this clear. Usually we would use a dash like
`for secret key-, public key-, and certificate-objects.` but this does not look right either....

Probably so?
`for secret key objects, public key objects, and certificate objects.`

Comment (ifranzki) [reply man/man1/p11sak.1.in]: ok, already changed and force pushed.

**Review by jschmidb — APPROVED** [review]
looks good

**Pull request merged**